### PR TITLE
Partially implemented mimicry.

### PR DESCRIPTION
**Pull request by Electrolyzer**
Edge cases where other effects also change type may exhibit some unintended behavior (soak, forest's curse).

Streamlined on summon weather/terrain effects by adding an onTerrainChange and onWeatherChange call when switching in and removing the relevant onSummonEffect so that when coding an ability that activates on specific terrain or weather, you only need to implement the onTerrainChange or onWeatherChange attribute. Added a getWeather() function to arena.ts to simplify it (identical to the preceding getTerrain() function).

### CODE REVIEW

File: src/data/ability.ts [hunk 1]
```diff
@@ -2126,6 +2126,43 @@ export class PostTerrainChangeAddBattlerTagAttr extends PostTerrainChangeAbAttr
...
+      pokemon.summonData.types = null;
+      pokemon.updateInfo();
+      pokemon.scene.queueMessage(getPokemonMessage(pokemon, ` transformed\nback into its original type!`));
+      return true;
+    }
+
+    pokemon.summonData.types = [this.terrainToTypeMap[terrain]]
+    pokemon.updateInfo();
+    pokemon.scene.queueMessage(getPokemonMessage(pokemon, ` transformed\ninto the ${Utils.toReadableString(Type[pokemon.getTypes()[0]])} type!`));
+    return true;
```
Comment (bennybroseph) [inline src/data/ability.ts:2162]: Could we get some comments explaining this function?

**Review by bennybroseph — CHANGES_REQUESTED** [review]

File: src/data/ability.ts (hunk 1 above)
Comment (Electrolyzer) [reply src/data/ability.ts:2162]: Added comments. Let me know if there is anything else or something in particular I should explain more. Thank you!

Comment (Electrolyzer) [reply src/data/ability.ts:2162]: (If the question was about terrainToTypeMap that was already explained above in the comment on the constructor)

**Comment by flx-sta** [comment]
No action within more than a month. 
Considered stale.

Feel free to re-open it when you are still working on it/want to work on it again

**Pull request closed without merging**